### PR TITLE
Add models during runtime

### PR DESCRIPTION
**Pull request by diegoferigo**
As discussed in #44, Ignition Gazebo now has the capability of adding models during runtime. The only way to do it using upstream features is passing through the `UserCommands` plugin. This plugin contains a system that provides a service that receives commands that define new entities.

I thought that it would be convenient for us to directly access the ECM and perform operations with the same approach of `UserCommands` without passing through a service. Therefore, this PR introduces a new `ECMProvider` plugin to share the ECM with other parts of the code through a singleton.

With this new support, the `GazeboWrapper` can now operate directly on the ECM. This PR adds to the `GazeboWrapper` the capability of creating a removing models during runtime.

The python package has been updated accordingly.

pendulum_runtime

### CODE REVIEW

File: ignition/src/ECMSingleton.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+class ECMSingleton::Impl
+{
+public:
+    mutable std::mutex mutex;
+    ignition::gazebo::EventManager* eventMgr = nullptr;
+    ignition::gazebo::EntityComponentManager* ecm = nullptr;
+};
+
+ECMSingleton::ECMSingleton()
+    : pImpl{new Impl(), [](Impl* impl) { delete impl; }}
```
Comment (traversaro) [inline ignition/src/ECMSingleton.cpp]: This seems to be exactly the default deleter, there is any reason to pass it explicitly?

Comment (diegoferigo) [reply ignition/src/ECMSingleton.cpp]: I remember I wrote the reason behind this workaround somewhere (maybe blockfactory / wb-toolbox) but I cannot find it anymore. It is a common pattern I typically use.

If I recall, the problem is due to the use of a unique pointer for the Impl object. The header has a forward declaration of the class, and the implementation is in the cpp file. In order to know how to free the memory for the Impl object (handled by the unique pointer), the forward declaration is not enough.

The workaround is initializing the unique pointer with a custom deleter (so you tell the compiler that you will handle it properly somehow), and you specify the deleter in the cpp.

See as an example:

https://github.com/robotology/gym-ignition/blob/aa91b25a23c44305b9a62de74a3c0f7eeeed4006/ignition/include/gympp/gazebo/RobotSingleton.h#L28-L29

Comment (traversaro) [reply ignition/src/ECMSingleton.cpp]: Ah, I see, I guess the reason is explained in https://www.fluentcpp.com/2017/09/22/make-pimpl-using-unique_ptr/ (where they use another technique to workaround this issue).

Comment (diegoferigo) [reply ignition/src/ECMSingleton.cpp]: I remember I read and tried that solution, but still there were cases where it didn't work. I found that doing that plus the custom deleter is a consistent solution that always works (and it does not add any extra line of code).

**Comment by diegoferigo** [comment]
Anyway, the ignition robotics framework doesn't really like clang. Downstream code compiles fine, but there's no way to make python tests pass, they always segfault :man_shrugging:

Let's stick to gcc7 and the recommended gcc8.

**Comment by diegoferigo** [comment]
Edit: Apparently also gcc8 [segfaults similarly](https://github.com/robotology/gym-ignition/runs/249778560#step:8:7099). Maybe there's something related to the compiler version used to build the packages of the ppa (which should be the default compiler in ubuntu bionic, gcc7).

**Comment by traversaro** [comment]
This week I will have limited time, so if someone else in the lab wants to review this/you want to merge this without my review, please go ahead.

**Comment by diegoferigo** [comment]
@traversaro Ok thanks for the notification. Since this is a rather involved PR I think I'm going to merge it as it is. I'd like to release soon a first beta release. I worked a lot on the tests and hopefully everything should be fine.

If you're interested either now or in the future to understand how to deal with the ECM, this PR is quite descriptive.

**Pull request merged**